### DISCUSSION TITLE
Collection property initializer (dumping IEnumerable into readonly collection property)

### DISCUSSION THREAD

**Discussion by GiottoVerducci** [dotnet/csharplang#5875] [Language Ideas]
In order to avoid null reference exception for collections, it's useful to have read-only properties initialized with empty collections, for instance:

```
public class Foo
{
	public int Bar { get; set; }
	public List<int> Doh { get; } = new List<int>();
	public List<Foo> Stuff { get; } = new List<Foo>();
}
```

but the initialization is not very elegant:

```
var foo = new Foo { Bar = 14; }
foo.AddRange(Enumerable.Range(1, 10));
foo.AddRange(new [] { otherFoo });
```

Adding a constructor with optional parameters requires a lot of boiler plate code:

```
public class Foo
{
	constructor Foo(IEnumerable<int> doh = null, IEnumerable<Foo> stuff = null)
	{
		if(doh != null) Doh.AddRange(doh);
		if(stuff != null) Stuff.AddRange(stuff);
	}
	...
}
```

at which point it completely defeats the purpose of initializers.

It could be nice to have a "dump into" operator with the "<=" syntax:

```
var foo = new Foo 
{ 
	Bar = 14,
	Doh <= Enumerable.Range(1, 10),
	Stuff <= new [] { otherFoo },
}
```

This would call AddRange (or Add in a loop) when initializing the property.

**Comment by GiottoVerducci** [comment]
(as a side-note, it would also work with Span<T>: ``` Doh <= mySpan.Slice(1, 4),```)

**Comment by HaloFour** [comment]
The language already supports this:

```cs
var foo = new Foo() {
    Bar = 1234,
    Stuff = { new Foo() },
    Doh = { 1, 2, 3 }
};
```

**Reply by canton7** [reply]
The one thing it doesn't support is `foo.AddRange(someOtherEnumerable)`. I think I've seen another proposal for this somewhere...

**Reply by HaloFour** [reply]
Right, it's a normal collection initializer and they only call `Add`.  There are separate proposals to support `AddRange` out there and theoretically if those are ever implemented it would "just work" with the existing syntax.

**Reply by GiottoVerducci** [reply]
There would be a syntax issue even with the support of AddRange since ```Doh = Enumerable.Range(1, 10),``` would likely return a compilation error (cannot assign a value to a readonly property)

That's where the new operator <= would be useful. 

Note that it doesn't need the support of AddRange because Add can be called in a loop, just as it does when using the ```Doh = { 1, 2, 3 }``` syntax.

**Comment by GiottoVerducci** [comment]
I found a workaround, but it's ugly: https://stackoverflow.com/questions/53431659/object-initializer-for-readonly-properties-in-c-sharp

Add a ```public static void Add<T>(this ICollection<T> collection, IEnumerable<T> items)``` extension method and write ```Doh = { someOtherEnumerable }```

The dump operator would be cleaner as it wouldn't pollute ICollection and wouldn't require the extra braces.

**Reply by GiottoVerducci** [reply]
If the "<="  would create ambiguity with the comparison operator, it could be replaced with "=<"
